KAFKA-3613: Consolidate TumblingWindows and HoppingWindows into TimeWindows

miguno opened the pull request:
This PR includes the same code as https://github.com/apache/kafka/pull/1261 but is rebased on latest trunk.

Commit:
KAFKA-3613: new

miguno commented:
@guozhangwang feel free to review (/cc @enothereska and @mjsax ).

guozhangwang commented on streams/examples/src/main/java/org/apache/kafka/streams/examples/pageview/PageViewTypedDemo.java:
```diff
@@ -160,7 +160,7 @@ public PageViewByRegion apply(PageView view, UserProfile profile) {
-                .countByKey(HoppingWindows.of("GeoPageViewsWindow").with(7 * 24 * 60 * 60 * 1000), Serdes.String())
+                .countByKey(TimeWindows.of("GeoPageViewsWindow", 7 * 24 * 60 * 60 * 1000L).shiftedBy(1000), Serdes.String())
                 // TODO: we can merge ths toStream().map(...) with a single toStream(...)
                 .toStream()
```
This is out of the scope of this PR but may just be OK do to together: since now we have added the toStream(map) function to KTable, we can actually get rid of this TODO.

guozhangwang commented on streams/src/main/java/org/apache/kafka/streams/kstream/UnlimitedWindows.java:
```diff
@@ -50,23 +50,30 @@ public UnlimitedWindows startOn(long start) {
-        // always return the single unlimited window
-
-        // we cannot use Collections.singleMap since it does not support remove() call
+        // Always return the single unlimited window.
```
For consistency we do not capitalize in-function comments.

Commit:
Fix time-based semantics of UnlimitedWindows

- Properly compute windows for a given timestamp (honoring window start time)
- Add unit tests/spec, notably to forbid negative window start times

guozhangwang commented:
LGTM overall. Just a couple of minor comments.

Commit:
Use diamond operator

Commit:
Use consistent javadoc grammar

guozhangwang commented:
@miguno Could you address the second comment above? No further reviews needed after that.

mjsax commented on streams/src/main/java/org/apache/kafka/streams/kstream/JoinWindows.java:
```diff
@@ -17,15 +17,14 @@
 import java.util.Map;
 
 /**
  * The window specifications used for joins.
```
Should we extend the JavaDoc? Or should this be done in #1250 ?

miguno replied:
I was waiting for https://github.com/apache/kafka/pull/1250 and intentionally didn't touch javadocs in this PR where I could avoid it.

mjsax commented on streams/src/main/java/org/apache/kafka/streams/kstream/TimeWindows.java:
```diff
@@ -0,0 +1,128 @@
+import java.util.Map;
+
+/**
+ * The time-based window specifications used for aggregations.
```
Should be more precise. See discussion about `HoppingWindow` and `TumblindWindow` in #1250

miguno replied:
As I said above, I intentionally didn't touch that because your PR at https://github.com/apache/kafka/pull/1250 is coming up.

mjsax commented on streams/src/main/java/org/apache/kafka/streams/kstream/TimeWindows.java:
```diff
@@ -0,0 +1,128 @@
+     * The hop size's effective time unit is determined by the semantics of the topology's
+     * configured {@link org.apache.kafka.streams.processor.TimestampExtractor}.
+     */
+    public final long hop;
```
I would prefer `advance` instead of `hop`.

As we discussed, `slide` should not be used to avoid confusion with sliding windows and `offset` would overload Kafka's  usage of this term.

enothereska replied:
This will probably be subjective, but I'm ok with "hop" for now.

mjsax commented on streams/src/main/java/org/apache/kafka/streams/kstream/TimeWindows.java:
```diff
@@ -0,0 +1,128 @@
+
+    /**
+     * Returns a window definition with the given window size, and with the hop size being equal to
+     * the window size.  Think: [N * size, N * size + size), with N denoting the N-th window.
```
Remove double blank.

mjsax commented on streams/src/main/java/org/apache/kafka/streams/kstream/TimeWindows.java:
```diff
@@ -0,0 +1,128 @@
+     * the window size.  Think: [N * size, N * size + size), with N denoting the N-th window.
+     *
+     * This provides the semantics of tumbling windows, which are fixed-sized, gap-less,
+     * non-overlapping windows.  Tumbling windows are a specialization of hopping windows.
```
Remove double blank.

mjsax commented on streams/src/main/java/org/apache/kafka/streams/kstream/TimeWindows.java:
```diff
@@ -0,0 +1,128 @@
+     *            topology's configured {@link org.apache.kafka.streams.processor.TimestampExtractor}.
+     * @return a new window definition
+     */
+    public TimeWindows shiftedBy(long hop) {
```
I would prefer `advanceBy(long advance)`

enothereska replied:
@mjsax makes a good point here. His recommendation was `advance` - `advanceBy.` If we stick with `hop,` perhaps we should have `hopBy`?

mjsax replied:
`hop` vs `advance` is subjective :) I am fine with `hop`, too; it's called a `HoppingWindow` after all. `advance` is just a term that is quite common in literature so I am somewhat used to it.

miguno replied:
I prefer `advanceBy` over `shiftedBy`, too.

What about the field though -- keep it `hop` or also change it to `advance`?  Which would you find more intuitive and/or consistent with prior art?

miguno replied:
Tentatively, I'd prefer `advanceBy(hop)`.

> `advanceBy`: Advances the window by the given hop size [...]

vs

> `advanceBy`: Advances the window by the given advance size [...]

That is, I find "advancing by advance" to sound a bit weird, and "advancing by [some other term, perhaps "hop"]" be a bit better.

mjsax replied:
I personally prefer consistent naming, ie, both `hop` or both `advance`.

mjsax replied:
You could phrase it differently: "Move the window by the given advance size." or "Advance the window by the given timespan." (or similiar)

miguno replied:
I updated the code by renaming shiftedBy/hop to advanceBy/advance.  See the latest PR for details.

mjsax commented on streams/src/main/java/org/apache/kafka/streams/kstream/TimeWindows.java:
```diff
@@ -0,0 +1,128 @@
+     *
+     * @param timestamp the timestamp
+     * @return a map of (windowStartTimestamp, window) entries
+     */
```
You can remove this JavaDoc -> is documents in Parent-Class via #1250

miguno replied:
Ack.

Commit:
Remove JavaDoc of windowsFor(), already covered by parent class

Commit:
Rename shiftedBy/hop to advanceBy/advance

mjsax commented on streams/src/main/java/org/apache/kafka/streams/kstream/TimeWindows.java:
```diff
@@ -35,27 +35,29 @@
-     * The hop size's effective time unit is determined by the semantics of the topology's
-     * configured {@link org.apache.kafka.streams.processor.TimestampExtractor}.
+     * The size of the window's advance interval, i.e. by how much a window moves forward relative
+     * to the previous one.  The interval's effective time unit is determined by the semantics of
```
Remove double blank.

mjsax commented on streams/src/main/java/org/apache/kafka/streams/kstream/TimeWindows.java:
```diff
@@ -35,27 +35,29 @@
-     * Returns a window definition with the given window size, and with the hop size being equal to
-     * the window size.  Think: [N * size, N * size + size), with N denoting the N-th window.
+     * Returns a window definition with the given window size, and with the advance interval being
+     * equal to the window size.  Think: [N * size, N * size + size), with N denoting the N-th
```
Remove double blank.

mjsax commented:
+1

guozhangwang commented:
LGTM. Merged to trunk.
